### PR TITLE
[MASSEMBLY-986] Minimal default Manifest configuration of jar archiver should be respected

### PR DESCRIPTION
**Pull request by slawekjaranowski**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MASSEMBLY) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MASSEMBLY-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MASSEMBLY-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java [hunk 1]
```diff
@@ -293,10 +294,7 @@ protected Archiver createArchiver(
         final List<FileSelector> extraSelectors = new ArrayList<>();
         final List<ArchiveFinalizer> extraFinalizers = new ArrayList<>();
         if (archiver instanceof JarArchiver) {
-            if (mergeManifestMode != null) {
-                ((JarArchiver) archiver)
-                        .setFilesetmanifest(JarArchiver.FilesetManifestConfig.valueOf(mergeManifestMode));
-            }
+            configureJarArchiver((JarArchiver) archiver, mergeManifestMode, configSource.getJarArchiveConfiguration());
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:297]: I would expect that this applies to all JAR derived formats? WAR, EAR, CAR, etc?

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java [hunk 2]
```diff
@@ -351,6 +349,23 @@ protected Archiver createArchiver(
...
+        if (mergeManifestMode != null) {
+            archiver.setFilesetmanifest(JarArchiver.FilesetManifestConfig.valueOf(mergeManifestMode));
+        }
+
+        archiver.setMinimalDefaultManifest(true);
+
+        if (configuration != null) {
+            archiver.setCompress(configuration.isCompress());
+            archiver.setForced(configuration.isForced());
+            archiver.setIndex(configuration.isIndex());
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:364]: I am confused, where do these come from?

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java [hunk 3]
```diff
@@ -351,6 +349,23 @@ protected Archiver createArchiver(
...
     }
 
+    private void configureJarArchiver(
+            JarArchiver archiver, String mergeManifestMode, MavenArchiveConfiguration configuration) {
+
+        if (mergeManifestMode != null) {
+            archiver.setFilesetmanifest(JarArchiver.FilesetManifestConfig.valueOf(mergeManifestMode));
+        }
+
+        archiver.setMinimalDefaultManifest(true);
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: This will set it?

**Comment by michael-o** [comment]
I am still not fully conviced here. For me the gain is questionable or I don't understand it completely. I'd like to hear @hboutemy's opinion on this.

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 3 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: yes - exactly

Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: So this leaves the user no option to include it?

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 2 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:364]: It comes from
https://github.com/apache/maven-archiver/blob/master/src/main/java/org/apache/maven/archiver/MavenArchiver.java#L576-L593

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 1 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:297]: yes - currently ear, rar, war 
https://codehaus-plexus.github.io/plexus-archiver/apidocs/org/codehaus/plexus/archiver/jar/JarArchiver.html

I would like to extends another types, eg for car we only have a UnArchiver
https://codehaus-plexus.github.io/plexus-archiver/apidocs/org/codehaus/plexus/archiver/car/package-summary.html

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 2 above)
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:364]: But these aren't related to this change, no? I am still searching for the connection between the issue title and this hunk.

Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:364]: Ok, I did it by the way - can be in separate issue 😄

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 3 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: The same is in Maven Archiver.

**Comment by hboutemy** [comment]
I don't understand what issue we're trying to solve
I don't understand the title of the issue
The only aspect I understand in the Jira issue description is that "it may break Reproducible Builds", but I don't see how: I never have RB broken by something like that (because Plexus Archiver component version is reproducible)

please explain what the problem is (theoretical or in real life) that is being tried to solve

**Comment by slawekjaranowski** [comment]
@hboutemy can you update your comment - is it more understandable for you now?

**Comment by hboutemy** [comment]
IIUC, what you are trying to achieve is to avoid Reproducible Builds output to change when changing `plexus-archiver` version (like proposed by dependabot in #119

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 3 above)
Comment (hboutemy) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: what if mergeManifestMode is set to `skip`?
I find it surprising to have `true` hardcoded

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 2 above)
Comment (hboutemy) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:364]: does this have to be checked against mergeManifestmode or not?

**Review by hboutemy — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 2 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:364]: As I wrote code was copied from maven-archiver, for this change I will remove it

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 3 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: > what if mergeManifestMode is set to `skip`? 

I don't test such ...

>I find it surprising to have `true` hardcoded

copied from maven-archived 😄

Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:359]: default value for `mergeManifestMode` is `skip`

 `setMinimalDefaultManifest` has only effect on code: https://github.com/codehaus-plexus/plexus-archiver/blob/40b1b12e849ef62021e9d9787add3db7f34267af/src/main/java/org/codehaus/plexus/archiver/jar/Manifest.java#L767

If we omit `addDefaultEntries` or set it with `true` - maven-archiver will add own Manifest 

So IMHO can be hardcoded

**Comment by slawekjaranowski** [comment]
Now PR contains only required changes to achieve goal.

File: src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java (hunk 3 above)
Comment (hboutemy) [reply src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java:357]: I definitively don't understand when all this is triggered, but it seems inoffensive, let's go...

**Review by hboutemy — APPROVED** [review]
LGTM

**Comment by jira-importer** [comment]
Resolve #1185

**Pull request merged**